test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/SignatureTrustAndValidityVerificationProviderTests.cs

```diff
@@ -329,6 +330,244 @@ public SignatureTrustAndValidityVerificationProviderTests(SigningTestFixture fix
+            {
+                var signedPackagePath = await SignedArchiveTestUtility.AuthorSignPackageAsync(
+                    testCertificate,
+                    nupkg,
+                    dir,
+                    timestampService.Url);
+
+                var countersignedPackagePath = await SignedArchiveTestUtility.RepositorySignPackageAsync(repoTestCertificate, signedPackagePath, dir, new Uri("https://v3serviceIndex.test/api/index.json"), timestampService.Url);
+
+                var verifier = new PackageSignatureVerifier(_trustProviders);
+                using (var packageReader = new PackageArchiveReader(countersignedPackagePath))
+                {
+                    // Act
+                    var result = await verifier.VerifySignaturesAsync(packageReader, settings, CancellationToken.None);
+                    var resultsWithErrors = result.Results.Where(r => r.GetErrorIssues().Any());
+
+                    // Assert
+                    result.Valid.Should().BeTrue();
+                    resultsWithErrors.Count().Should().Be(0);
+                }
+            }
+        }
+
+        [CIOnlyFact]
+        public async Task VerifySignaturesAsync_WithExpiredPrimarySignature_ValidCountersignature_AndPrimarySignatureValidAtCountersignTime_Succeeds()
+        {
+            // Arrange
+            var nupkg = new SimpleTestPackageContext();
+            var timestampService = await _testFixture.GetDefaultTrustedTimestampServiceAsync();
+            var settings = new SignedPackageVerifierSettings(
+                allowUnsigned: false,
+                allowIllegal: false,
+                allowUntrusted: false,
+                allowIgnoreTimestamp: true,
+                allowMultipleTimestamps: true,
+                allowNoTimestamp: true,
+                allowUnknownRevocation: true,
+                alwaysVerifyCountersignature: true,
+                allowNoClientCertificateList: true,
+                allowNoRepositoryCertificateList: true);
+            var certExpiry = new Stopwatch();
+
+            using (var dir = TestDirectory.Create())
+            using (var willExpireCert = new X509Certificate2(_testFixture.TrustedTestCertificateWillExpireIn5Seconds.Source.Cert))
+            using (var repoTestCertificate = new X509Certificate2(_trustedTestCert.Source.Cert))
+            {
+                certExpiry.Start();
+
+                var signedPackagePath = await SignedArchiveTestUtility.AuthorSignPackageAsync(
+                    willExpireCert,
+                    nupkg,
+                    dir);
+
+                var countersignedPackagePath = await SignedArchiveTestUtility.RepositorySignPackageAsync(repoTestCertificate, signedPackagePath, dir, new Uri("https://v3serviceIndex.test/api/index.json"), timestampService.Url);
+
+                // Wait for cert to expire
+                while (certExpiry.ElapsedMilliseconds < 5000)
+                {
+                    Thread.Sleep(100);
```

Comment:
You don't need a stopwatch.  You can just check current time against `willExpireCert.NotAfter`.  This feedback applies to other tests that use this certificate.